chain/events/filter/resolver.go

```diff
@@ -0,0 +1,74 @@
+package filter
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"lukechampine.com/blake3"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type (
+	// ActorResolver resolves an address to its actor ID at a given TipSet, or the latest TipSet if nil.
+	ActorResolver func(context.Context, address.Address, *types.TipSet) (abi.ActorID, error)
+
+	cachedActorResolver struct {
+		delegate       ActorResolver
+		cache          *expirable.LRU[string, cachedActorResolution]
+		cacheNilTipSet bool
+	}
+	cachedActorResolution struct {
+		result abi.ActorID
+		err    error
+	}
+)
+
+var cacheHasherPool = sync.Pool{
+	New: func() any {
+		return blake3.New(32, nil)
+	},
+}
+
+func NewCachedActorResolver(delegate ActorResolver, size int, expiry time.Duration, cacheNilTipSet bool) ActorResolver {
```

Comment:
In terms of caching, I believe we will get a bigger speed up by keeping the same state tree around while resolving addresses within a single tipset.

Reply:
But this probably is not a bottleneck.

Reply:
I can imagine this being helpful for the case of a public node operator serving event lookups for popular event types and addresses. It's probably not going to help much for general use but should scale in a multi-user situation.